### ISSUE TITLE
N(<string>, 30) and sympify: implement a decimal precision kwarg?

### ISSUE THREAD

**Issue by cbm755** [sympy/sympy#13425]
Passing a string to `N` is not explicitly documented, but it does work.  The implementation is `sympify(<input>).evalf(n)`.  Should we change `N` to `sympify(<input>, evaluate=False).evalf(n)`?  This improves accuracy in some very specific cases.

This came up downstream https://savannah.gnu.org/bugs/?51748

- - - - -

Here's the background.  The current code generally DTRT, for example:
````
>>> N('sqrt(2)', 30)
1.41421356237309504880168872421
>>> _**2
2.00000000000000000000000000000
````

In general if you put floats in the string, its not going to possible to get more than 15 digits of accuracy and you shouldn't be doing that.  But people do it anyway and then think its a bug:
````
>>> N('2**0.5', 30)
1.41421356237309514547462185874
>>> _**2
2.00000000000000027343234630648
````
But it is possible to get full precision in this case:
````
>>> sympify('2**0.5', evaluate=False).evalf(30)
1.41421356237309504880168872421
>>> _**2
2.00000000000000000000000000000
````

1.  In the current implementation `'2**0.5'` is evaluated to a Float with default precision.  Of course, this cannot be magically re-evaluated in higher precision.
2.  Deferring the evaluation works better in this specific case b/c the float `0.5` is exact in binary.
3.  Deferring will not help at all with `'2**0.1'` because `0.1` is not an integer power of 2 so will be inexact in binary.
4.  I worry this might make some other calculation worse (due to additional rounding error), but I have no concrete example.

**Comment by asmeurer** [comment]
Passing strings to anything other than sympify is not really supported, but I don't see an issue with fixing just N() for this, since it's clear that it could help. However, there could be side effects for evaluate=False, since that also prevents evaluation of symbols expressions (like `N('x - x', 30)`).

**Comment by asmeurer** [comment]
Maybe sympify() should gain a flag to change the precision of automatic evaluation. I wonder how difficult that would be to achieve.

**Comment by cbm755** [comment]
I also worried about side effects, so maybe the `evaluate=False` thing is indeed a bad idea.

Re: changing `sympify`: this sounds reasonable, maybe something like `sympify('2*0.1', default_float_dps=30)`.  l poked around; looks like `sympy.parsing.sympy_parser.auto_number` is where the `Float` is.  So this *might* be an easy fix, just need to ensure the new kwarg gets to `auto_number`.

Note: `sympify('0.12345678901234567890123456789')` already uses a high precision automatically (b/c Float does so).  So the default for `default_float_dps` should be `None` rather than 16).

**Comment by yashkgp** [comment]
Can I get more info on this?

**Comment by cbm755** [comment]
So first of all, I'm not *sure* its "Easy to Fix", I only say it might be...

1.  Try `srepr(sympify('2*0.1'))`.  Note result has double precision (16 digits, 53 bits).

2.  Try `Float(sympify('2*0.1'), 32)`.  Note result has higher precision but is half-full of garbage.  Think about and understand why approach will never work.

3.  Implement `sympify('2*0.1'), default_float_dps=32)` which uses `Float(<str>, 32)` at the time the Float is first created.

**Comment by avishrivastava11** [comment]
@cbm755 @asmeurer  can I work on this issue?
Is it still open?

**Comment by asmeurer** [comment]
I believe it is still open. 

Another possible solution to this is for `N` to use `parse_expr`, and to add a flag that gets passed through to `auto_number` that sets the precision on the `Float` token. The way the tokenize transformations are designed this isn't particularly easy to do, though. Probably the simplest way would be for it to be its own transformation, which gets added after `auto_number`.

**Comment by avishrivastava11** [comment]
then I'll start working on it...

**Comment by asmeurer** [comment]
I'm not really sure what fix should be applied here. I'm partly thinking it should just be "won't fix", as string inputs to things other than `sympify` isn't really supported. But maybe the token transformer idea that adds the precision to `Float` calls could be useful, even if it isn't used directly by `N`.

**Comment by avishrivastava11** [comment]
@cbm755 
>@avishrivastava11 sorry for the misleading "easy-to-fix" label. In general, if you're read over an issue, sunk a couple hours into the code a bit and still don't understand, then its probably not the right place to start! Having said all that, do feel free to play with #14824 if you're interested in this topic.

Thanks for the comment. Do you mind if I make a PR using some ideas and code from your PR #14824 (you mentioned that the PR was only for discussion, and not for merging). I think I can solve this issue once and for all. Do you mind?

**Comment by asmeurer** [comment]
@avishrivastava11 you might want to discuss what you plan to do here first, to avoid wasting time implementing something that we would end up rejecting. I'm of the opinion that very little should be done here, as most of the proposed changes add a lot of complexity for a very marginal benefit.

**Comment by cbm755** [comment]
> Do you mind if I make a PR using some ideas and code from your PR ...

To answer this question in general: a good workflow is to checkout my branch, and make additional commits (on top of my existing commits).  This ensures appropriate attribution.

**Comment by avishrivastava11** [comment]
@cbm755 @asmeurer  I'll do that. Thanks.

**Comment by smichr** [comment]
> there could be side effects for evaluate=False, since that also prevents evaluation of symbols expressions (like `N('x - x', 30)`).

The evalf will release the expression and 0 will be returned. Test added in #16780

**Comment by smichr** [comment]
If one wants arbitrary precision it is best to sympify with `rational=True` and not use `N` to sympify and evaluate:
```python
>>> S('2**.5', rational=True)
sqrt(2)
```
 The partial fix that I added to #16780 will only allow exact base-2 decimals but can't help the non-exact values:
```
>>> (N('pi*.1', 22)*10-pi).n()
1.74393421862262e-16
>>> (N('pi/10', 22)*10-pi).n()
-2.11259981339749e-23
>>> (S('pi/10').n(22)*10-pi).n()
-2.11259981339749e-23
```
